File: frontend/src/lib/mlBridge.ts
import { execFile } from 'node:child_process';
import * as path from 'path';
import { promisify } from 'node:util';
import * as fs from 'node:fs';

const execFileAsync = promisify(execFile);

type PythonCandidate = {
  bin: string;
  prefixArgs: string[];
};

function candidatePythonBins(): PythonCandidate[] {
  // Repo layout assumption:
  // - this runs in `frontend/`
  // - project root is `../` (NutriScan)
  const frontendDir = path.resolve(process.cwd());
  const projectRoot = path.resolve(frontendDir, '..');

  const winVenv = path.join(projectRoot, '.venv', 'Scripts', 'python.exe');
  const unixVenv = path.join(projectRoot, '.venv', 'bin', 'python3');

  const candidates: PythonCandidate[] = [];

  // 1) Explicit env var (supports full path)
  if (process.env.PYTHON_BIN) candidates.push({ bin: process.env.PYTHON_BIN, prefixArgs: [] });

  // 2) Project venv (best on Windows)
  if (fs.existsSync(winVenv)) candidates.push({ bin: winVenv, prefixArgs: [] });

  // 3) Project venv (unix)
  if (fs.existsSync(unixVenv)) candidates.push({ bin: unixVenv, prefixArgs: [] });

  // 4) Windows Python launcher (avoids Microsoft Store "python"/"python3" alias issues)
  if (process.platform === 'win32') {
    candidates.push({ bin: 'py', prefixArgs: ['-3'] });
  }

  // 5) System Python fallbacks
  candidates.push({ bin: 'python', prefixArgs: [] });
  candidates.push({ bin: 'python3', prefixArgs: [] });

  return candidates;
}

function scriptPath(): string {
  return path.resolve(process.cwd(), '..', 'ml', 'ingredient_model', 'frontend_infer.py');
}

export async function runMlBridge(args: string[]): Promise<any> {
  const script = scriptPath();
  const bins = candidatePythonBins();

  let lastError: unknown = null;

  for (const cand of bins) {
    try {
      const { stdout, stderr } = await execFileAsync(cand.bin, [...cand.prefixArgs, script, ...args], {
        cwd: path.resolve(process.cwd(), '..'),
        timeout: 300000,
        windowsHide: true,
        maxBuffer: 1024 * 1024 * 8,
        env: { ...process.env, PYTHONWARNINGS: 'ignore' },
      });

      const text = stdout.trim();
      if (!text) {
        throw new Error('ML bridge returned empty output');
      }

      let parsed;
      try {
        parsed = JSON.parse(text);
      } catch (e) {
        console.error('[ML Bridge] JSON parse error. stderr:', stderr);
        console.error('[ML Bridge] stdout:', text);
        throw e;
      }

      if (parsed && typeof parsed === 'object' && 'error' in parsed) {
        // Error from ML bridge is expected (e.g., product not found) - still return it
        return parsed;
      }
      return parsed;
    } catch (err) {
      lastError = err;
      console.error(`[ML Bridge] Failed with ${cand.bin}: ${err}`);
    }
  }

  throw new Error(`ML bridge execution failed: ${String(lastError)}`);
}

File: backend/services/ml_bridge.py
import json
import os
import subprocess
from pathlib import Path
from typing import Any


BASE_DIR = Path(__file__).resolve().parents[2]
SCRIPT_PATH = BASE_DIR / "ml" / "ingredient_model" / "frontend_infer.py"


def _python_candidates() -> list[tuple[str, list[str]]]:
    candidates: list[tuple[str, list[str]]] = []

    env_bin = os.getenv("PYTHON_BIN", "").strip()
    if env_bin:
        candidates.append((env_bin, []))

    local_win_venv = BASE_DIR / ".venv" / "Scripts" / "python.exe"
    local_unix_venv = BASE_DIR / ".venv" / "bin" / "python3"
    workspace_win_venv = BASE_DIR.parent / ".venv" / "Scripts" / "python.exe"
    workspace_unix_venv = BASE_DIR.parent / ".venv" / "bin" / "python3"

    for candidate in [local_win_venv, local_unix_venv, workspace_win_venv, workspace_unix_venv]:
        if candidate.exists():
            candidates.append((str(candidate), []))

    if os.name == "nt":
        candidates.append(("py", ["-3"]))

    candidates.append(("python", []))
    candidates.append(("python3", []))
    return candidates


def run_ml_bridge(args: list[str], timeout_seconds: int = 300) -> Any:
    if not SCRIPT_PATH.exists():
        raise RuntimeError(f"ML bridge script not found at {SCRIPT_PATH}")

    last_error = "unknown error"
    for python_bin, prefix_args in _python_candidates():
        cmd = [python_bin, *prefix_args, str(SCRIPT_PATH), *args]
        try:
            result = subprocess.run(
                cmd,
                cwd=str(BASE_DIR),
                capture_output=True,
                text=True,
                timeout=timeout_seconds,
                check=False,
                env={**os.environ, "PYTHONWARNINGS": "ignore"},
            )
        except Exception as exc:
            last_error = f"{python_bin}: {exc}"
            continue

        if result.returncode != 0:
            last_error = f"{python_bin}: {result.stderr.strip() or result.stdout.strip() or 'non-zero exit code'}"
            continue

        payload = result.stdout.strip()
        if not payload:
            last_error = f"{python_bin}: empty output"
            continue

        try:
            return json.loads(payload)
        except json.JSONDecodeError as exc:
            last_error = f"{python_bin}: invalid JSON output ({exc})"
            continue

    raise RuntimeError(f"ML bridge execution failed: {last_error}")